x/stake/keeper/slash.go

```diff
@@ -192,44 +200,52 @@ func (k Keeper) slashUnbondingDelegation(ctx sdk.Context, unbondingDelegation ty
 // insufficient stake remaining)
 // nolint: unparam
 func (k Keeper) slashRedelegation(ctx sdk.Context, validator types.Validator, redelegation types.Redelegation,
-	infractionHeight int64, slashFactor sdk.Dec) (slashAmount sdk.Int) {
+	infractionHeight int64, slashFactor sdk.Dec) (totalSlashAmount sdk.Int) {
 
 	now := ctx.BlockHeader().Time
+	totalSlashAmount = sdk.ZeroInt()
 
-	// If redelegation started before this height, stake didn't contribute to infraction
-	if redelegation.CreationHeight < infractionHeight {
-		return sdk.ZeroInt()
-	}
-
-	if redelegation.MinTime.Before(now) {
-		// Redelegation no longer eligible for slashing, skip it
-		// TODO Delete it automatically?
-		return sdk.ZeroInt()
-	}
+	// perform slashing on all entries within the redelegation
+	for i, entry := range redelegation.Entries {
 
-	// Calculate slash amount proportional to stake contributing to infraction
-	slashAmountDec := slashFactor.MulInt(redelegation.InitialBalance.Amount)
-	slashAmount = slashAmountDec.TruncateInt()
+		// If redelegation started before this height, stake didn't contribute to infraction
+		if entry.CreationHeight < infractionHeight {
+			continue
+		}
 
-	// Don't slash more tokens than held
-	// Possible since the redelegation may already
-	// have been slashed, and slash amounts are calculated
-	// according to stake held at time of infraction
-	redelegationSlashAmount := sdk.MinInt(slashAmount, redelegation.Balance.Amount)
+		if entry.IsMature(now) {
+			// Redelegation no longer eligible for slashing, skip it
+			// TODO Delete it automatically?
```

Comment:
It should be deleted at the end of the block. Clearer to keep the deletion logic in one place I think.